driver/vhci/vhci_write.c

```diff
@@ -351,21 +351,28 @@ process_write_irp(pusbip_vpdo_dev_t vpdo, PIRP irp)
 	}
 
 	status = process_urb_res(urbr, hdr);
-
-	if (urbr->irp != NULL) {
-		IoSetCancelRoutine(urbr->irp, NULL);
-		urbr->irp->IoStatus.Status = status;
-
-		/* it seems windows client usb driver will think
-		 * IoCompleteRequest is running at DISPATCH_LEVEL
-		 * so without this it will change IRQL sometimes,
-		 * and introduce to a dead of my userspace program
-		 */
-		KeRaiseIrql(DISPATCH_LEVEL, &oldirql);
-		IoCompleteRequest(urbr->irp, IO_NO_INCREMENT);
-		KeLowerIrql(oldirql);
-	}
+	PIRP irp = urbr->irp;
 	free_urbr(urbr);
+	
+	if (irp != NULL) {
+		BOOLEAN valid_irp;
+		IoAcquireCancelSpinLock(&oldirql);
+		valid_irp = IoSetCancelRoutine(irp, NULL) != NULL;
+		IoReleaseCancelSpinLock(oldirql);
+		if (valid_irp) {
+			irp->IoStatus.Status = status;
+
+			/* it seems windows client usb driver will think
```

Comment:
@cezanne i must say that i tought it was weird that there is a need to raiseirql here, and had to see it myself - wow - looks like some ancient f-up in the microsoft stack

Reply:
IRQL raising code was a legacy. I remember it was very weird. I'll also investigate.